In [1]:
import nltk, os

os: 외부에서 파일을 불러올 때 사용하면 편리한 라이브러리. os는 우리가 흔히 쓰는 윈도우, 맥과 같은 운영체제를 일컬음.

In [2]:
s = 'a man in the house' # untokenized string
t = ['a', 'man', 'in', 'the', 'house'] # tokenized seqeuence of words as a list

s 는 아직 토큰화하지 않은 하나의 텍스트 문자열  
t 는 토큰화된 단어들로 이루어진 리스트

In [3]:
a = nltk.Text(t)
a

<Text: a man in the house...>

nltk.Text()
: A wrapper around a sequence of simple (string) tokens, which is intended to support initial exploration of texts (via the interactive console). 일반적인 리스트를 counting, concordancing, collocation등의 기능을 수행하는 nltk의 다양한 메서드들을 적용시킬 수 있는 Class인 Text클라스로 변환해주는 함수.

In [4]:
nltk.Text(s.split())

<Text: a man in the house...>

s.split() 문자열을 스페이스를 기준으로 구분하여 리스트로 만들면, 토큰화된 단어 리스트가 될 수 있다.

In [5]:
raw = open(os.getcwd()+r'/06_01.txt', encoding = 'utf8').read()

os.getcwd()는 현재 디렉토리의 주소를 가져옴
현재 디렉토리에 있는 06_01.txt 파일을 utf8으로 인코딩하여 열고(open), 그것을 문자열로 읽어서(read) 변수 raw에 지정.

In [6]:
print(type(raw))
print(len(raw))
print(raw[:75])

<class 'str'>
1154507
The Project Gutenberg EBook of Crime and Punishment, by Fyodor Dostoevsky




read()라는 함수로 텍스트 파일을 하나의 문자열로 불러왔으니 변수 raw의 자료형은 str(문자열)이다.  
len(raw)는 변수 raw에 할당된 문자열의 문자수다.  
raw[:75]는 문자열의 0번째부터 74번째 문자까지의 문자열을 반환하는 슬라이싱이다.

In [7]:
tokens = nltk.word_tokenize(raw)
print(type(tokens))
print(len(tokens))
print(tokens[:10])

<class 'list'>
257726
['The', 'Project', 'Gutenberg', 'EBook', 'of', 'Crime', 'and', 'Punishment', ',', 'by']


word_tokenize는 단어 단위로 토큰화해주는 함수. 입력값은 문자열, 출력값은 토큰들로 이루어진 리스트.

In [8]:
text = nltk.Text(tokens)
print(type(text))
print(text[:10])
text.collocations()


<class 'nltk.text.Text'>
['The', 'Project', 'Gutenberg', 'EBook', 'of', 'Crime', 'and', 'Punishment', ',', 'by']
Katerina Ivanovna; Pyotr Petrovitch; Pulcheria Alexandrovna; Avdotya
Romanovna; Rodion Romanovitch; Marfa Petrovna; Sofya Semyonovna; old
woman; Project Gutenberg-tm; Porfiry Petrovitch; Amalia Ivanovna;
great deal; young man; Nikodim Fomitch; Ilya Petrovitch; Project
Gutenberg; Andrey Semyonovitch; Hay Market; Dmitri Prokofitch; Good
heavens


nltk.Text() 함수의 역할은 주어진 입력의 자료형을 nltk.text.Text으로 만들어줌. nltk 전용 variable로 만들어줌.
nltk전용 variable로 만들어주면, nltk만의 함수를 적용할 수 있다.  
collocations()는 같이 자주 나오는 단어 묶음을 알려줌.

In [9]:
text.concordance('great', 79, 25)


Displaying 25 of 25 matches:
and were more frequent in periods of great strain . In 1859 he was allowed to r
ndency and of late she had read with great interest a book she got through Mr. 
 the bosom of her family ... . And a great deal more ... . Quite excusable , si
that you had heard that Dounia had a great deal to put up with in the Svidrigra
g will she has . Dounia can endure a great deal and even in the most difficult 
 that letter she reproached him with great heat and indignation for the basenes
putation ; they had seen and known a great deal more than Mr. Svidrigailov had 
n other people ’ s . In my opinion a great deal , a very great deal of all this
 In my opinion a great deal , a very great deal of all this was unnecessary ; b
 . He is a very busy man and is in a great hurry to get to Petersburg , so that
 me that , though he is not a man of great education , he is clever and seems t
 very well . Of course , there is no great love either on his side , or on hers
tted the ma

concordance()는 입력으로 주어진 단어가 등장하는 모든 문맥을 보여줌. 79는 가로로 79개의 문자를 보여주고, 25는 세로로 25줄을 보여줌.

In [10]:
raw = nltk.corpus.gutenberg.raw('melville-moby_dick.txt')
fdist = nltk.FreqDist(raw)
print(fdist.most_common(5))

fdist = nltk.FreqDist(ch.lower() for ch in raw if ch.isalpha())
print(fdist.most_common(5))
[char for (char, count) in fdist.most_common(5)]

[(' ', 198098), ('e', 115855), ('t', 85539), ('a', 75266), ('o', 68338)]
[('e', 117092), ('t', 87996), ('a', 77916), ('o', 69326), ('n', 65617)]


['e', 't', 'a', 'o', 'n']

nltk 라이브러리는 여러 예제 데이터들이 들어가있음.  
FreqDist()함수는 입력으로 들어간 것의 빈도 분포를 보여줌.  
most_common(n)은 그 빈도분포에서 가장 빈도수가 높은 상위 n개를 보여줌.  

[ch.lower() for ch in raw if ch.isalpha()]
문자열 raw의 각각의 문자(character)에 대해서 그 문자가 알파벳이면 그 문자를 소문자화시켜 리스트의 원소로 둔다.  

nltk.FreqDist(ch.lower() for ch in raw if ch.isalpha())  
위 리스트의 원소들에 대해서 빈도분포를 만든다.  

print(fdist.most_common(5))  
빈도분포에서 빈도수 상위 5개의 리스트(문자-빈도수 pair들로 이루어진 리스트)를 화면 출력하라.  

** [char for (char, count) in fdist.most_common(5)]에서 for loop는 몇번 도는가? :5번

In [11]:
nltk.corpus.words.words('en')[-20:-1]

['zymosis',
 'zymosterol',
 'zymosthenic',
 'zymotechnic',
 'zymotechnical',
 'zymotechnics',
 'zymotechny',
 'zymotic',
 'zymotically',
 'zymotize',
 'zymotoxic',
 'zymurgy',
 'Zyrenian',
 'Zyrian',
 'Zyryan',
 'zythem',
 'Zythia',
 'zythum',
 'Zyzomys']

nltk.corpus.words 에서 영어('en') 단어리스트를 가져오고, 리스트의 끝에서 20번째부터 끝에서 첫번째 원소들을 리스트 형태로 가져온다. 즉, 영어 단어들 중에서 뒤에서 20번째에서 1번째 단어들을 리스트로 가져오라는 말이다.

In [12]:
len(nltk.corpus.words.words('en'))

235886

nltk.corpus.words에 있는 영어 단어들의 개수는 235886개다.